### PR TITLE
New Rule: Duo Admin Lockout

### PR DESCRIPTION
**Pull request by andrea-youwakim**
### Background

<High level overview here>

### Changes

* newly written duo rule to alert on admin lockout

### Testing

* updated test results: 
Screen Shot 2022-12-21 at 12 30 10 PM

### CODE REVIEW

File: rules/duo_rules/duo_admin_lockout.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+def rule(event):
+    # Return True to match the log event and trigger an alert.
+    return event.get("action", "") == "admin_lockout"
+
+
+def title(event):
+    # If no 'dedup' function is defined, the return value
+    # of this method will act as deduplication string.
+    return (
+        f"Duo User [{event.get('username', '<NO_USER_FOUND>')}] is "
```
Comment (jpl5280) [inline rules/duo_rules/duo_admin_lockout.py:10]:
```suggestion
        f"Duo Admin [{event.get('username', '<NO_USER_FOUND>')}] is "
```

File: rules/duo_rules/duo_admin_lockout.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    # Return True to match the log event and trigger an alert.
+    return event.get("action", "") == "admin_lockout"
+
+
+def title(event):
+    # If no 'dedup' function is defined, the return value
+    # of this method will act as deduplication string.
+    return (
+        f"Duo Admin [{event.get('username', '<NO_USER_FOUND>')}] is "
+        f"temporarily locked out due to too many passcode attempts."
```
Comment (jpl5280) [inline rules/duo_rules/duo_admin_lockout.py:11]: I'm not sure the description is static (aka `Admin temporarily locked out due to too many passcode attempts.` may only be 1 of many reasons'). Suggestion: 

pseudocode"

`f"locked out. Reason: event.get('description').get('message')"`

File: rules/duo_rules/duo_admin_lockout.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        description: '{"bypass": "", "count": 1, "valid_secs": 3600, "auto_generated": true, "remaining_uses": 1, "user_id": "D12345", "bypass_code_ids": ["A12345"]}'
+        isotimestamp: "2022-12-14 21:17:39"
+        object: target@example.io
+        timestamp: "2022-12-14 21:17:39"
+        username: Homer Simpson
+      Name: Bypass Create
+DedupPeriodMinutes: 60
+LogTypes:
+    - Duo.Administrator
+    - Duo.Authentication
```
Comment (jpl5280) [inline rules/duo_rules/duo_admin_lockout.yml:29]: i dont think we need this one, in this case?

File: rules/duo_rules/duo_admin_lockout.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+def rule(event):
+    # Return True to match the log event and trigger an alert.
+    return event.get("action", "") == "admin_lockout"
+
+
+def title(event):
+    # If no 'dedup' function is defined, the return value
+    # of this method will act as deduplication string.
+    return (
+        f"Duo User [{event.get('username', '<NO_USER_FOUND>')}] is "
```
Comment (jpl5280) [inline rules/duo_rules/duo_admin_lockout.py:10]:
```suggestion
        f"Duo Admin [{event.get('username', '<NO_USER_FOUND>')}] is "
```

**Review by jpl5280 — APPROVED** [review]

File: rules/duo_rules/duo_admin_lockout.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    return event.get("action", "") == "admin_lockout"
+
+
+def title(event):
+    # If no 'dedup' function is defined, the return value
+    # of this method will act as deduplication string.
+    return (
+        f"Duo Admin [{event.get('username', '<NO_USER_FOUND>')}] is "
+        f"locked out. Reason: "
+        f"[{event.get('description', {}).get('message','<NO_MESSAGE_FOUND>')}]."
```
Comment (edyesed) [inline rules/duo_rules/duo_admin_lockout.py:12]: I expect that this is aways returning `<NO_MESSAGE_FOUND>` from the test suite. `event.description` is a string which needs to be `json.loads()`'ed to understand that there is a `message` attribute in there.

**Review by edyesed — CHANGES_REQUESTED** [review]

**Review by edyesed — APPROVED** [review]
duo detection for admin lockout

**Pull request merged**